### PR TITLE
Fix ConvexSet move operations

### LINKED ISSUE #19310: Invalid move implementations for ConvexSet subclasses
Reading the code, I'm pretty sure that a moved-from ConvexSet subclass does not meet the "valid but indeterminate state" requirement.  The subclass's fields will be empty (e.g., the Point's `x_`), but the superclass does not annotate `reset_after_move` on the ambient dimension. The `cloner_` is probably null as well.

### PR DESCRIPTION
**Pull request by jwnimmer-tri**
Requires #19321.  Closes #19310.  Towards #19279.

- - -
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19311)

### CODE REVIEW

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 23 of 23 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19311)* status: needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @jwnimmer-tri)

**Comment by jwnimmer-tri** [comment]
\CC @RussTedrake FYI, or in case you'd like to feature-review.

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19311)* status: needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @jwnimmer-tri)

**Comment by jwnimmer-tri** [comment]
On second look, there are still several lingering bugs related to zero-dimensional set handling.  I've converted back to "Draft" while I work on those.

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 26 of 26 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19311)* status: needs platform reviewer assigned, needs at least two assigned reviewers, labeled "do not merge" (waiting on @jwnimmer-tri)

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19311)* status: needs platform reviewer assigned, needs at least two assigned reviewers, labeled "do not merge" (waiting on @jwnimmer-tri)

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19311)* status: 1 unresolved discussion, needs platform reviewer assigned, needs at least two assigned reviewers, labeled "do not merge"

___
*[`geometry/optimization/convex_set.cc` line 47 at r6](https://reviewable.io/reviews/RobotLocomotion/drake/19311#-NUN0yTy1XAZmnst4ity:-NUN0yTz64J_FmNJZ78g:bh9huxj) ([raw file](https://github.com/RobotLocomotion/drake/blob/2fc07f737dfd54e6c402424dc89bde52f7ac5f86/geometry/optimization/convex_set.cc#L47)):*
> ```cpp
>     const Eigen::Ref<const solvers::VectorXDecisionVariable>& x,
>     const symbolic::Variable& t) const {
>   DRAKE_THROW_UNLESS(ambient_dimension() > 0);
> ```

Working

Some unit tests are hitting this.

Either they are incorrectly formulated (in which case we should fix the test), or they really do need to add this vacuously false  constraint, in which case we need to figure out how to add a vacuously false constraint to MathematicalProgram without it throwing an exception.

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19311)* status: needs platform reviewer assigned, needs at least two assigned reviewers, labeled "do not merge"

___
*[`geometry/optimization/convex_set.cc` line 47 at r6](https://reviewable.io/reviews/RobotLocomotion/drake/19311#-NUN0yTy1XAZmnst4ity:-NUNkY9q3dKyINCQCfZZ:brc3fxy) ([raw file](https://github.com/RobotLocomotion/drake/blob/2fc07f737dfd54e6c402424dc89bde52f7ac5f86/geometry/optimization/convex_set.cc#L47)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Working

Some unit tests are hitting this.

Either they are incorrectly formulated (in which case we should fix the test), or they really do need to add this vacuously false  constraint, in which case we need to figure out how to add a vacuously false constraint to MathematicalProgram without it throwing an exception.
</blockquote></details>

Actually, moving this discussion to #19321.

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 3 of 25 files at r7, 29 of 29 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19311)* status: LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers (waiting on @jwnimmer-tri and @RussTedrake)

**Comment by RussTedrake** [comment]
*[`geometry/optimization/convex_set.h` line 244 at r8](https://reviewable.io/reviews/RobotLocomotion/drake/19311#-NXKkX5i5-UO3KkUL8vi:-NXKkX5i5-UO3KkUL8vj:bfobm3s) ([raw file](https://github.com/RobotLocomotion/drake/blob/ba3478aaa779190e1e73c5b82df4fe605d065ef8/geometry/optimization/convex_set.h#L244)):*
> ```objc
> 
>  private:
>   reset_after_move<int> ambient_dimension_;
> ```

in your comments about threading, you said "readability/maintainability of the code may trump other concerns".  In my view, you are making something that should be simple look very complicated.  (here an in the lines above). I think it's unlikely that developers will get this right out of the box.  I've just now read the reset_after_move docs (i was not aware)... but don't know if that is enough for me to justify this complexity?

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19311)* status: 1 unresolved discussion, LGTM missing from assignee RussTedrake(platform), needs at least two assigned reviewers

___
*[`geometry/optimization/convex_set.h` line 244 at r8](https://reviewable.io/reviews/RobotLocomotion/drake/19311#-NXKkX5i5-UO3KkUL8vi:-NXL42NG4XvBTieO6wPE:b-jomc6j) ([raw file](https://github.com/RobotLocomotion/drake/blob/ba3478aaa779190e1e73c5b82df4fe605d065ef8/geometry/optimization/convex_set.h#L244)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

in your comments about threading, you said "readability/maintainability of the code may trump other concerns".  In my view, you are making something that should be simple look very complicated.  (here an in the lines above). I think it's unlikely that developers will get this right out of the box.  I've just now read the reset_after_move docs (i was not aware)... but don't know if that is enough for me to justify this complexity?
</blockquote></details>

It didn't read as complicated at all to me, but let's run with the idea anyway.

If we assume that this is too difficult as-is, I'm not sure where you're going with the thought process.  Maybe we could add more documentation nearby here, to better explain?  Something like:

```c++
  // When this object is moved from (via the move constructor or move
  // assignment operator), the reset_after_move wrapper will reset the
  // ambient dimension to zero. This is important because subclasses
  // that are composite sets (e.g., MinkowskiSum) will have their
  // child sets moved away, leaving them with no dimensions to report.
```

The bottom line is that the move constructors are all broken on master.  Unless we have a fix that ends up with this same effect (however we spell it), the only other alternative is to remove (`= delete`) all of the move operators on `ConvexSet` and all of its subclasses, leaving only the copy operators.  Is that a better fix than this approach?

**Review by RussTedrake — COMMENTED** [review]
:lgtm:

> Reviewed 3 of 25 files at r7, 29 of 29 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19311#-:-NXLMcKf5TMhUsoJ5tVx:bg7igk9)* status: 1 unresolved discussion, needs at least two assigned reviewers

___
*[`geometry/optimization/convex_set.h` line 244 at r8](https://reviewable.io/reviews/RobotLocomotion/drake/19311#-NXKkX5i5-UO3KkUL8vi:-NXLIV1z55xcEe685_VW:bp3ao57) ([raw file](https://github.com/RobotLocomotion/drake/blob/ba3478aaa779190e1e73c5b82df4fe605d065ef8/geometry/optimization/convex_set.h#L244)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

It didn't read as complicated at all to me, but let's run with the idea anyway.

If we assume that this is too difficult as-is, I'm not sure where you're going with the thought process.  Maybe we could add more documentation nearby here, to better explain?  Something like:

```c++
  // When this object is moved from (via the move constructor or move
  // assignment operator), the reset_after_move wrapper will reset the
  // ambient dimension to zero. This is important because subclasses
  // that are composite sets (e.g., MinkowskiSum) will have their
  // child sets moved away, leaving them with no dimensions to report.
```

The bottom line is that the move constructors are all broken on master.  Unless we have a fix that ends up with this same effect (however we spell it), the only other alternative is to remove (`= delete`) all of the move operators on `ConvexSet` and all of its subclasses, leaving only the copy operators.  Is that a better fix than this approach?
</blockquote></details>

The last line of the comment helps me understand the goal.  I initially failed to see the danger in having the ambient dimension have a non-zero value sticking around.

It's complicated because it suddenly asks the develops to understand a lot more about move semantics, etc.  And one has to go and read the reset_after_move header file to learn more.  We have important contributors to the ConvexSet classes who are still coming up to speed on the more nuanced aspects of C++, and I don't want to scare them all away.

I think adding the comment here provides sufficient justification for the added complexity.

**Review by sherm1 — COMMENTED** [review]
Platform :lgtm: with two suggestions

> Reviewed 2 of 25 files at r7, 27 of 29 files at r8, 3 of 3 files at r9.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19311#-:-NXMehP22BvJeQs6eENg:b-3fpmlv)* status: 2 unresolved discussions

___
*[`geometry/optimization/test/cartesian_product_test.cc` line 80 at r9](https://reviewable.io/reviews/RobotLocomotion/drake/19311#-NXMc0kh6HDQuvopElYf:-NXMc0kh6HDQuvopElYg:b-l4tr0j) ([raw file](https://github.com/RobotLocomotion/drake/blob/55f0809def762c3c408197f25a6f7d96aed27d5a/geometry/optimization/test/cartesian_product_test.cc#L80)):*
> ```cpp
>   // The old set is in a valid but unspecified state. For convenience we'll
>   // assert that it's empty, but that's not the only valid implementation,
>   // just the one we happen to currently use.
> ```

BTW I don't think it's necessary or helpful to talk here about other implementations we might have done. Consider removing the second sentence (META). (Also the current behavior seems like the obvious choice -- likely everyone will be happy with it forever.)
___
*[`geometry/optimization/test/spectrahedron_test.cc` line 230 at r9](https://reviewable.io/reviews/RobotLocomotion/drake/19311#-NXMeGzL079cUdF1tqiO:-NXMeGzL079cUdF1tqiP:b-vyrqf7) ([raw file](https://github.com/RobotLocomotion/drake/blob/55f0809def762c3c408197f25a6f7d96aed27d5a/geometry/optimization/test/spectrahedron_test.cc#L230)):*
> ```cpp
>   EXPECT_NO_THROW(dut.AddPointInSetConstraints(
>       &constraints, constraints.NewContinuousVariables<6>()));
>   EXPECT_EQ(constraints.positive_semidefinite_constraints().size(), 1);
> ```

BTW not clear to me what this additional test is accomplishing -- can you add a comment explaining?

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 3 of 3 files at r9, 1 of 1 files at r10, 1 of 1 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19311)* status: :shipit: complete! all discussions resolved, LGTM from assignees sherm1(platform),RussTedrake(platform)

___
*[`geometry/optimization/test/cartesian_product_test.cc` line 80 at r9](https://reviewable.io/reviews/RobotLocomotion/drake/19311#-NXMc0kh6HDQuvopElYf:-NXMf3CJ97-l7K7IkjVJ:b-p18htk) ([raw file](https://github.com/RobotLocomotion/drake/blob/55f0809def762c3c408197f25a6f7d96aed27d5a/geometry/optimization/test/cartesian_product_test.cc#L80)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW I don't think it's necessary or helpful to talk here about other implementations we might have done. Consider removing the second sentence (META). (Also the current behavior seems like the obvious choice -- likely everyone will be happy with it forever.)
</blockquote></details>

In general I think it's important for unit tests to make it clear when they are asserting something beyond the documented contract of the API.

Would there be a less confusing phrasing you can think of?
___
*[`geometry/optimization/test/spectrahedron_test.cc` line 230 at r9](https://reviewable.io/reviews/RobotLocomotion/drake/19311#-NXMeGzL079cUdF1tqiO:-NXMg5NA-mx_G3naqXCM:b-896fix) ([raw file](https://github.com/RobotLocomotion/drake/blob/55f0809def762c3c408197f25a6f7d96aed27d5a/geometry/optimization/test/spectrahedron_test.cc#L230)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW not clear to me what this additional test is accomplishing -- can you add a comment explaining?
</blockquote></details>

Done.

**Comment by sherm1** [comment]
*[`geometry/optimization/test/cartesian_product_test.cc` line 80 at r9](https://reviewable.io/reviews/RobotLocomotion/drake/19311#-NXMc0kh6HDQuvopElYf:-NXMl-W78X6kSoZwH-78:b-7chaor) ([raw file](https://github.com/RobotLocomotion/drake/blob/55f0809def762c3c408197f25a6f7d96aed27d5a/geometry/optimization/test/cartesian_product_test.cc#L80)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

In general I think it's important for unit tests to make it clear when they are asserting something beyond the documented contract of the API.

Would there be a less confusing phrasing you can think of?
</blockquote></details>

IMO the first sentence achieved that adequately. Several of the new tests had just the first sentence for the same condition; I thought that was fine.

**Comment by jwnimmer-tri** [comment]
*[`geometry/optimization/test/cartesian_product_test.cc` line 80 at r9](https://reviewable.io/reviews/RobotLocomotion/drake/19311#-NXMc0kh6HDQuvopElYf:-NXMlwp77QeXJlqvk3Uh:b-i6fi2x) ([raw file](https://github.com/RobotLocomotion/drake/blob/55f0809def762c3c408197f25a6f7d96aed27d5a/geometry/optimization/test/cartesian_product_test.cc#L80)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

IMO the first sentence achieved that adequately. Several of the new tests had just the first sentence for the same condition; I thought that was fine.
</blockquote></details>

Right, those were the tests that only tested the "valid but unspecified" predicate.  In the tests that assert "empty" instead of "unspecified", we have the extra text.  I think I'm OK leaving it.
